spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.spark.dispatcher;
+
+import static org.opensearch.sql.spark.data.constants.SparkConstants.DRIVER_ENV_ASSUME_ROLE_ARN_KEY;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.EXECUTOR_ENV_ASSUME_ROLE_ARN_KEY;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_DELEGATE_CATALOG;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_INDEX_STORE_AUTH_KEY;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_INDEX_STORE_AWSREGION_KEY;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_INDEX_STORE_HOST_KEY;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_INDEX_STORE_PORT_KEY;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.FLINT_INDEX_STORE_SCHEME_KEY;
+import static org.opensearch.sql.spark.data.constants.SparkConstants.HIVE_METASTORE_GLUE_ARN_KEY;
+
+import com.amazonaws.services.emrserverless.model.GetJobRunResult;
+import com.amazonaws.services.emrserverless.model.JobRunState;
+import java.net.URI;
+import java.net.URISyntaxException;
+import lombok.AllArgsConstructor;
+import org.json.JSONObject;
+import org.opensearch.sql.datasource.DataSourceService;
+import org.opensearch.sql.datasource.model.DataSourceMetadata;
+import org.opensearch.sql.spark.client.EmrServerlessClient;
+import org.opensearch.sql.spark.jobs.model.S3GlueSparkSubmitParameters;
+import org.opensearch.sql.spark.response.JobExecutionResponseReader;
+
+/** This class takes care of understanding query and dispatching job query to emr serverless. */
+@AllArgsConstructor
+public class SparkQueryDispatcher {
+
+  private EmrServerlessClient emrServerlessClient;
+
+  private DataSourceService dataSourceService;
+
+  private JobExecutionResponseReader jobExecutionResponseReader;
+
+  public String dispatch(String applicationId, String query, String executionRoleARN) {
+    String datasourceName = getDataSourceName();
+    try {
+      return emrServerlessClient.startJobRun(
+          query,
+          "flint-opensearch-query",
+          applicationId,
+          executionRoleARN,
+          constructSparkParameters(datasourceName));
+    } catch (URISyntaxException e) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Bad URI in indexstore configuration of the : %s datasoure.", datasourceName));
+    }
+  }
+
+  public JSONObject getQueryResponse(String applicationId, String jobId) {
+    GetJobRunResult getJobRunResult = emrServerlessClient.getJobRunResult(applicationId, jobId);
+    JSONObject result = new JSONObject();
+    if (getJobRunResult.getJobRun().getState().equals(JobRunState.SUCCESS.toString())) {
```

Comment:
add todo. fetch from query_result index first. if no result, fetch from emr-s.

Reply:
Hey what if the emr-s is still in the process of writing the result and customer queries. There is a chance, I can get half-baked results. But if we can live with that, I can switch.

Reply:
Added TODO.